### PR TITLE
Log feature gates on gardener-apiserver startup

### LINKED ISSUE #4579: Log feature gates on gardener-apiserver startup
**How to categorize this issue?**

/area usability
/kind enhancement
/priority 3

**What would you like to be added**:
Currently gardenlet, gardener-controller-manager and gardener-scheduler log the feature gates on startup:

https://github.com/gardener/gardener/blob/d9d79a4bd84b5e5679369817c7031358e202f7dd/cmd/gardenlet/app/gardenlet.go#L241

https://github.com/gardener/gardener/blob/d9d79a4bd84b5e5679369817c7031358e202f7dd/cmd/gardener-controller-manager/app/gardener_controller_manager.go#L206

https://github.com/gardener/gardener/blob/d9d79a4bd84b5e5679369817c7031358e202f7dd/cmd/gardener-scheduler/app/gardener_scheduler.go#L118

It would be nice if gardener-apiserver can also log its feature gates on startup.

### PR DESCRIPTION
**Pull request by ialidzhikov**
/area ops-productivity usability
/kind enhancement

Fixes #4579

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: cmd/gardener-apiserver/app/gardener_apiserver.go [hunk 1]
```diff
@@ -247,6 +248,11 @@ func (o *Options) config(kubeAPIServerConfig *rest.Config, kubeClient *kubernete
 
 // Run runs gardener-apiserver with the given Options.
 func (o *Options) Run(ctx context.Context) error {
+	logger := logger.NewLogger("", "")
+	logger.Info("Starting Gardener API server...")
+	logger.Infof("Version: %+v", version.Get())
+	logger.Infof("Feature Gates: %s", utilfeature.DefaultFeatureGate)
```
Comment (vpnachev) [inline cmd/gardener-apiserver/app/gardener_apiserver.go:254]: If I start it with no feature gate flag set, it logs
```
{"level":"info","msg":"Feature Gates: ","time":"2021-09-21T18:33:19+03:00"}
```

Can it be made to print all feature gates and their enablement status? E.g. print `DisallowKubeconfigRotationForShootInDeletion=true` even when it is not explicitly enabled via CLI flag?

Comment (stoyanr) [reply cmd/gardener-apiserver/app/gardener_apiserver.go:254]: On the other hand, I think GCM, `gardenlet` etc. would similarly not log any feature gates not explicitly enabled / disabled, so we change it for the apiserver, we should change it everywhere.

**Review by stoyanr — COMMENTED** [review]
/lgtm except for the question already raised

File: cmd/gardener-apiserver/app/gardener_apiserver.go (hunk 1 above)
Comment (ialidzhikov) [reply cmd/gardener-apiserver/app/gardener_apiserver.go:254]: Valid point. The reason is that the featureGate String implementation considers only explicitly enabled features and their values and does not consider known features and their default values.

Ref https://github.com/kubernetes/kubernetes/blob/092fbfbf53427de67cac1e9fa54aaa09a28371d7/staging/src/k8s.io/component-base/featuregate/feature_gate.go#L247-L255

I am even considering that the String() func can be implemented in such way because in K8s there are a lot of known features (more that 100) and the output itself would "explode" with such high amount of feature gates.

So that's why I personally wouldn't invest into adapting the upstream func.

So bottom line, we have to treat these logged feature gates as the ones that are explicitly configured for the component (and not as the ones that also contain the default known feature gates and their values - this is what is actually used by the `Enabled` func itself at runtime). 😅

Comment (stoyanr) [reply cmd/gardener-apiserver/app/gardener_apiserver.go:254]: OK, then lets' leave this as it is.

**Review by stoyanr — APPROVED** [review]
/lgtm

**Review by rfranzke — APPROVED** [review]
/lgtm

**Review by vpnachev — APPROVED** [review]
/lgtm
